Paper: Search for Extraterrestrial Intelligence: GPU Accelerated E.T. Search with TurboSETI

luigifcruz opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@cbcunc<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@m-maciejewski<!--end-reviewers-list-->

luigifcruz committed:
first commit

luigifcruz committed:
Add authors.

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

luigifcruz committed:
Format references.

luigifcruz committed:
Fix reference name.

luigifcruz committed:
Fix table width.

luigifcruz committed:
Add reference to tables.

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"luigifcruz-2022_lfc"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpxeg8q024'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: Luigi_Cruz\n*** Warning: /tmp/tmpxeg8q024/publisher/../output/Luigi_Cruz/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/luigifcruz-2022_lfc.pdf", "build_status"=>"success", "build_timestamp"=>"29/05 02:05"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @cbcunc as editor

scoobies commented:
Assigned! @cbcunc is now the editor

cbcunc commented:
Editor confirms paper builds.

deniederhut commented:
@scoobies add @m-maciejewski to reviewers

scoobies commented:
@m-maciejewski added to the reviewers list!

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

cbcunc commented:
@m-maciejewski First review is due June 16.

m-maciejewski commented:
@cbcunc I'm on it!

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+:bibliography: reference
+
+------------------------------------------------------------------------------------
+Search for Extraterrestrial Intelligence: GPU Accelerated E.T. Search with TurboSETI
```
I'd suggest to expand the E.T. abbreviation as it might not be clear for the broader audience. Once the abbreviation is expanded, the title seems to repeat itself. How about: "GPU Accelerated Search for Extraterrestrial Intelligence with TurboSETI Package"?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+
+.. class:: abstract
+
+A common technique adopted by the SETI community is monitoring electromagnetic radiation for signs of extraterrestrial technosignatures using ground-based radio observatories. 
```
Please spell-out SETI as it appears for the first time.

m-maciejewski replied:
Does E.T. stand for "extraterrestrial technosignatures"?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+A common technique adopted by the SETI community is monitoring electromagnetic radiation for signs of extraterrestrial technosignatures using ground-based radio observatories. 
+The analysis is made using a Python-based software called TurboSETI to detect narrowband drifting signals inside the recordings that can mean a technosignature (E.T.).
+The data stream generated by a telescope can easily reach the rate of terabits per second. 
+Our goal was to improve the processing speeds by writing a GPU-accelerated backend in addition to the original CPU-based implementation of the de-doppler algorithm.
```
Please consider adding a sentence explaining why would one need the de-doppler algorithm

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The Search for Extraterrestrial Intelligence (SETI) is a broad term utilized to describe the effort of locating any scientific proof of past or present technology that originated beyond the bounds of Earth. 
+SETI can be performed in a plethora of ways: either actively by deploying orbiters and rovers around planets/moons within the solar system, or passively by either searching for biosignatures in exoplanet atmospheres or “listening” to technologically-capable extraterrestrial civilizations.
+One of the most common techniques adopted by the SETI community is monitoring electromagnetic radiation for narrowband signs of technosignatures using ground-based radio observatories.
+This search can be performed in equipment primarily built for this task, like the Allen Telescope Array (California, USA), renting observation time, or in the background while the primary user is conducting other observations. 
```
Please split this sentence in two as it is hard to read in one go. E.g.,
This search can be performed in equipment primarily built for this task, like the Allen Telescope Array (California, USA).
The study is carried out either by renting observation time, or in the background while the primary user is conducting other observations.

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+This search can be performed in equipment primarily built for this task, like the Allen Telescope Array (California, USA), renting observation time, or in the background while the primary user is conducting other observations. 
+Other radio-observatories useful for this search include the MeerKAT Telescope (Northern Cape, South Africa), Green Bank Telescope (West Virginia, USA), and the Parkes Telescope (New South Wales, Australia).
+The operation of a radio-telescope is similar to an optical telescope.
+Instead of using optics to concentrate light into an optical sensor, a radio-telescope operates by concentrating electromagnetic waves into an antenna using a large reflective structure called a “dish”.
```
Could you provide a reference to a "dish"?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The operation of a radio-telescope is similar to an optical telescope.
+Instead of using optics to concentrate light into an optical sensor, a radio-telescope operates by concentrating electromagnetic waves into an antenna using a large reflective structure called a “dish”.
+The interaction between the metallic antenna and the electromagnetic wave generates a faint electrical current.
+This effect is then quantized by an analog-to-digital converter as RAW voltages and transmitted to a processing logic to extract useful information from it.
```
What do you mean by RAW voltage?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+Instead of using optics to concentrate light into an optical sensor, a radio-telescope operates by concentrating electromagnetic waves into an antenna using a large reflective structure called a “dish”.
+The interaction between the metallic antenna and the electromagnetic wave generates a faint electrical current.
+This effect is then quantized by an analog-to-digital converter as RAW voltages and transmitted to a processing logic to extract useful information from it.
+The data stream generated by a radio telescope can easily reach the rate of terabits per second.
```
What is the reaseon for data rate in the order of terabits per second? Explaining that would help the reader in diving deeper into your work.

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The current workflow utilized by the Breakthrough Listen, the largest scientific research program aimed at finding evidence of extraterrestrial intelligence, consists in pre-processing and storing the incoming data as frequency-time binary files (:cite:`lebofsky_et_al`) in persistent storage for later analysis. 
+This post-analysis is made possible using a Python-based software called TurboSETI (:cite:`enriquez_2019`) to detect narrowband signals that could be drifting in frequency owing to the relative radial velocity between the observer on earth, and the transmitter.
+Although being an offline operation, the processing speed of TurboSETI is in lockstep with the amount of science we can extract from each observation. Each filterbank file ingested by TurboSETI is often on the order of a few hundreds of gigabytes. To process data efficiently, the program uses Numpy for near machine-level performance. 
+To measure a potential signal's drift rate, TurboSETI uses a de-doppler algorithm to align the frequency axis according to a pre-set drift rate. Another algorithm called “hitsearch” is then utilized to identify any signal present in the recorded spectrum. 
```
Do you have a reference to the "hitsearch"?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The data stream generated by a radio telescope can easily reach the rate of terabits per second.
+The current workflow utilized by the Breakthrough Listen, the largest scientific research program aimed at finding evidence of extraterrestrial intelligence, consists in pre-processing and storing the incoming data as frequency-time binary files (:cite:`lebofsky_et_al`) in persistent storage for later analysis. 
+This post-analysis is made possible using a Python-based software called TurboSETI (:cite:`enriquez_2019`) to detect narrowband signals that could be drifting in frequency owing to the relative radial velocity between the observer on earth, and the transmitter.
+Although being an offline operation, the processing speed of TurboSETI is in lockstep with the amount of science we can extract from each observation. Each filterbank file ingested by TurboSETI is often on the order of a few hundreds of gigabytes. To process data efficiently, the program uses Numpy for near machine-level performance. 
```
Please try to avoid using jargon. I'd argue that one can't extract science from anything as it stands on its own ;-)
Instead, you may extract a meaningful insight, relevant information, etc.

m-maciejewski replied:
What do you mean by "near machine-level performance"?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+2. Approach
+-----------
+
+Multiple methods were utilized in this effort to write a GPU-accelerated backend and optimize the CPU implementation of TurboSETI. In this section, we enumerate them.
```
Multiple methods -> Three, seperate approaches/methods were utilized

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The original implementation of TurboSETI heavily depends on Numpy (:cite:`harris2020array`) for data processing. 
+To keep the number of modifications as low as possible, we implemented the GPU-accelerated backend using CuPy (:cite:`cupy_learningsys2017`). 
+This open-source library offers GPU acceleration backed by NVIDIA CUDA and AMD ROCm while using a Numpy style API. 
+This enabled us to share most of the code between the CPU and GPU-based implementations.
```
reuse most of the code?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+Some computationally heavy methods of the original CPU-based implementation of TurboSETI were written in Cython.
+This approach has disadvantages: the developer has to be familiar with Cython syntax to alter the code; the code requires additional logic to be compiled at installation time. 
+Consequently, it was decided to replace Cython with pure Python methods decorated with the Numba (:cite:`numba_2022`) accelerator. 
+By leveraging the power of LLVM’s Just-In-Time compiler, Numba can compile Python code into assembly code as well as apply SIMD acceleration instructions to achieve near machine-level speeds.
```
Please spell out the abbreviations

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+2.2. Single-Precision Floating-Point
+++++++++++++++++++++++++++++++++++++
+
+The original implementation of the software handled the input data as double-precision floating-point numbers. This behavior would cause all the mathematical operations to take significantly longer to process because of the added precision. The ultimate precision of the output product is inherently limited by the precision of the original input data which in most cases is represented by an 8-bit  ...
```
added precision - augmented/extended precision

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+----------
+
+To test the speed improvements between implementations we used filterbank files from previous observations coming from different observatories. 
+Table :ref:`table:double-precision` indicates the processing times it took to process three different files in double-precision mode. 
```
Please mention that the double-precision mode on CPU is your reference.

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+    :label:`table:single-precision`
+
+    +-----------------------------------------------------------+
+    | Double-Precision (float64)                                |
```
Single-Precision?

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+
+To illustrate the processing time improvement, a single observation containing 105 GB of data was processed in 12 hours by the original CPU-based TurboSETI implementation on an i7-7700K Intel CPU, and just 1 hour and 45 minutes by the GPU-accelerated backend on a GTX 1070 Ti NVIDIA GPU.
+
+4. Conclusion
```
Please add  a comment on the outcome of your benchmark. What would be the recommended setup?

m-maciejewski commented:
Overall, a well-written abstract. Please take a look at the suggested comments. I'm open for discussing them in detail. I do hope, the review would improve the abstract!

deniederhut commented:
@m-maciejewski I'm not seeing any inline comments -- did you click the "submit review" button?

m-maciejewski commented:
@deniederhut You can see all my comments in this thread with `Pending` status. Alternatively, they are all available here: https://github.com/scipy-conference/scipy_proceedings/pull/722/files

luigifcruz commented:
Hi, @m-maciejewski! Thanks for reviewing our paper. We can't see your comments, I think you need to submit them for us to be able to see them.

m-maciejewski commented:
@luigifcruz what do you mean by submitting? Here is a pdf of this website, where you can immediately see my comments; please just scroll up this page
[review.pdf](https://github.com/scipy-conference/scipy_proceedings/files/8931400/review.pdf)
.

luigifcruz commented:
Hey @m-maciejewski! Yes, these comments are private to you. You need to submit the review to make it available to others. 

> Before you submit your review, your line comments are pending and only visible to you. You can edit pending comments anytime before you submit your review. To cancel a pending review, including all of its pending comments, scroll down to the end of the timeline on the Conversation tab, then click Cancel review.

You can find the documentation regarding this GitHub feature [here](https://docs.github.com/en/enterprise-server@3.1/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request).

deniederhut commented:
Yup so @m-maciejewski  right now your comments look like this:

<img width="1087" alt="Screen Shot 2022-06-17 at 9 12 18 PM" src="https://user-images.githubusercontent.com/6868396/174418627-6c1ccfde-59a8-4417-9943-f8c5754e9e9d.png">

you'll need to click Finish Review, then Submit Review, which looks like this

<img width="1091" alt="Screen Shot 2022-06-17 at 9 13 00 PM" src="https://user-images.githubusercontent.com/6868396/174418649-31272a9f-8cb5-4288-87ea-f03740001620.png">

Let us know if you need any help!

m-maciejewski commented:
I'm sorry for not submitting the review. I work daily on GitLab where this feature works differntly.

deniederhut commented:
No worries! Glad we got it figured out

deniederhut commented:
@scoobies mark pending comment

cbcunc commented on papers/Luigi_Cruz/reference.bib:
```diff
@@ -0,0 +1,71 @@
+  month        = {May} 
+}
+
+@inproceedings{cupy_learningsys2017,
```
Needs a DOI (proceeding DOI will do).

luigifcruz replied:
I couldn't find the DOI of this publication. Their repository ([CITATIONS.md](https://github.com/kmaehashi/cupy/blob/master/CITATION.bib)) also doesn't have a DOI.

cbcunc commented on papers/Luigi_Cruz/reference.bib:
```diff
@@ -0,0 +1,71 @@
+  pages        = {124505}
+}
+
+‌@article{enriquez_2019,
```
Needs DOI.

luigifcruz replied:
Added on c1b642d0422e3f864c5c5036cb24daa1b3b5bb9d. Thanks!

deniederhut commented:
Hi @luigifcruz ! Have you had a chance to respond to the reviewers comments?

luigifcruz committed:
Add DOI to citations.

luigifcruz committed:
Addressing @m-maciejewski comments

luigifcruz commented:
Thanks for the comments! I just pushed a new commit addressing the comments made by @cbcunc and @m-maciejewski. I hope the paper is now more informative.

deniederhut commented:
Awesome! @m-maciejewski do you feel this paper is now ready for inclusion in the proceedings?

cbcunc commented:
Reviewers, please give your final recommendations. The deadline for final review comments is tomorrow.

Authors, please respond to review comments with changes or comments which resolve the conversation. Final acceptance decisions are July 4.

m-maciejewski commented on papers/Luigi_Cruz/paper.rst:
```diff
@@ -0,0 +1,123 @@
+The data stream generated by a radio telescope can easily reach the rate of terabits per second because of the ultra-wide bandwidth radio spectrum.
+The current workflow utilized by the Breakthrough Listen, the largest scientific research program aimed at finding evidence of extraterrestrial intelligence, consists in pre-processing and storing the incoming data as frequency-time binary files (:cite:`lebofsky_et_al`) in persistent storage for later analysis. 
+This post-analysis is made possible using a Python-based software called TurboSETI (:cite:`enriquez_2019`) to detect narrowband signals that could be drifting in frequency owing to the relative radial velocity between the observer on earth, and the transmitter.
+Although being an offline operation, the processing speed of TurboSETI is in lockstep with the amount of science we can extract from each observation. Each voltage file ingested by TurboSETI is often on the order of a few hundreds of gigabytes. To process data efficiently without Python overhead, the program uses Numpy for near machine-level performance. 
```
"amount of science we can extract from each observation" does not read well. Please use a more appropriate phrase.

luigifcruz replied:
Thanks for the comment. I rewrote the phrase using clearer language.

m-maciejewski commented:
Dear authors, 
thank you for considering the comments.
I've just suggested a one tiny comment. Otherwise, the text reads very well - one can easily follow from one sentence to another. Well done!

luigifcruz committed:
Rewrite phrase.

cbcunc commented:
@scoobies mark ready
